README.adoc

```diff
@@ -70,16 +82,314 @@ $ bundle exec liquidoc -d _data/data-sample.yml -t _templates/liquid/tpl-sample.
+
+For standard-format files that have non-standard filenames (for example, `.js` for a JSON file), as well as for free-form files, you must declare a type explicitly.
+
+[source,yaml]
+.Example -- Explicitly instructing file type for YAML data source
+----
+compile:
+  data:
+    file: source_data_file.yaml
+    type: yml
+  builds:
+    - template: liquid_template.html
+      output: _output/output_file.html
+----
+
+This is a bit counterintuitive in this particular example, but it's the same idea if you you're stuck with a `.js` file that's actually just JSON.
+
+[source,yaml]
+.Example -- Instructing correct type for mislabeled JSON file
+----
+compile:
+  data:
+    file: source_data_file.js
+    type: json
+  builds:
+    - template: liquid_template.html
+      output: _output/output_file.html
+----
+
+Once LiquiDoc is knows the right file type, it will oarse the file into a Ruby hash data structure for further processing.
+
+==== CSV Data
+
+Data ingested from CSV files will use the first row as key names for columnar data in the subsequent rows, as shown below.
+
+.Example -- sample.csv showing header/key and value rows
+[source,csv]
+----
+name,description,default,required
+enabled,Whether project is active,,true
+timeout,The duration of a session (in seconds),300,false
+----
+
+The above source data, parsed as a CSV file, will yield an _array_.
+Each array item represents a row from the CSV file (except the first row).
+Each array item contains a _structure_, or what Ruby calls a _hash_.
+As represented in the CSV example above, if the structure contains more than one key-value pair (more than one “column” in the source), all such pairs will be siblings, not nested or hierarchical.
+
+.Example -- array derived from sample.csv, with values depicted
+[source,ruby]
+----
+data[0].name #=> enabled
+data[0].description #=> Whether project is active
+data[0].default #=> nil
+data[0].required #=> true
+data[1].name #=> timeout
+data[1].description #=> The duration of a session (in seconds)
+data[1].default #=> 300
+data[1].required #=> false
+----
+
+==== Free-form Data
+
+Free-form data can only be parsed using regex patterns -- otherwise LiquiDoc has no idea what to consider data and what to consider noise.
+
+Any file organized with one record per line may be consumed and parsed by LiquiDoc, provided you instruct the parser regarding what variables to extract from where.
+The parser will read each line individually, applying your regex pattern to extract data.
+
+[TIP]
+.Learn regular expressions
+If you're already familiar enough with regex, this note is not for you.
+If you deal with docs but are not a regex user, become one.
+I promise you will deem the initial hurdles worth surmounting.
+
+.Example -- sample.free free-form data source file
+----
+A_B A thing that *SnASFHE&"\|+1Dsaghf true
+G_H Some text for &hdf 1t`F false
+----
+
+.Example -- regular expression with named groups for variable generation
```

Comment:
Seems like it's required to use named capture groups?  In other words, do non-named capture groups work and fields are just numbered, or will they straight-up cause failures?

If named capture groups are required, you may want to bring that to more attention.  If not, having a numbered group in the example output would help

Reply:
If unnamed patterns will work, I haven't figured out how to express the variables in Liquid. I was hoping we wouldn't need to do named groups, since it adds lots of room for frustration and error. Hopefully someone will complain, and then I'll try to figure out how those number vars are carried and drop that requirement. I'll note it and maybe describe them a bit better for now.